vulnerabilities/overflow-underflow.md

```diff
@@ -1,73 +1,158 @@
+
+Consider the contract below which provides a method to get a swap quote by adding one to the input amount:
+
+```solidity
+interface IDexPair {
+    function getSwapQuoteUint128(uint128 amountToSwap) external view returns(uint128);
+}
+
+contract DexPair is IDexPair {
+    function getSwapQuoteUint128(uint128 amountToSwap) external view returns(uint128 outputTokens) {
+        assembly {
+            outputTokens := add(amountToSwap, 1)
+            if lt(outputTokens, amountToSwap) { revert(0, 0) }
+        }
+    }
+}
+```
+
+- The problem is that the `add` opcode always produces a 256-bit result. For `uint128` maximum value (`type(uint128).max`), this will not overflow in the 256-bit space but will overflow when treated as `uint128`
+- The overflow check inside assembly fails because it checks against 256-bit values.
+
 
 ### Use of unchecked code block:
-Performing arithmetic operations inside the unchecked block saves a lot of gas because it omits several checks and opcodes. But, some of these opcodes are used in default arithmetic operations in 0.8 to check for underflow/overflow.
+Performing arithmetic operations inside the unchecked block saves a lot of gas because it omits several checks and opcodes. But some of these opcodes are used in default arithmetic operations in 0.8 to check for underflow/overflow.
 
 ```solidity
 uint8 public a = 255;
 
 function uncheck() public{
 
   unchecked {
-      a++;  // overflow and reset to 0 without reverting
+      a++;      // overflow and reset to 0 without reverting
   }
 
 }
 ```
 The unchecked code block is only recommended if you are sure that there is no possible way for the arithmetic to overflow or underflow.
 
+
+### Mitigation Strategies
+
+1. Using `addmod`
+
+```solidity
+function getSwapQuoteUint128(uint128 amountToSwap) external view returns(uint128 outputTokens) {
+    assembly {
+        outputTokens := addmod(amountToSwap, 1, 340282366920938463463374607431768211455)
+        if lt(outputTokens, amountToSwap) { revert(0, 0) }
+    }
+}
+```
+
+The `addmod` operation confines the result within the `uint128` range.
+
+2. Post-assembly Check
+
+```solidity
+function getSwapQuoteUint128(uint128 amountToSwap) external view returns(uint128 outputTokens) {
+    assembly {
+        outputTokens := add(amountToSwap, 1)
+    }
+    require(outputTokens >= amountToSwap, "Overflow detected!");
+}
+```
+
+Perform an overflow check outside YUL using normal Solidity to ensure correct behavior for `uint128` as it will compare the 128-bit values.
+
+3. Add a manual check
+
+```solidity
+function getSwapQuote(uint256 amountToSwap) external view returns(uint256 outputTokens) {
+    assembly {
+        outputTokens := add(amountToSwap, 1)
+        if lt(outputTokens, amountToSwap) { revert(0, 0) }
+    }
+}
+```
+
+This check ensures that if the result is less than the input, it indicates an overflow.
+
+> [!IMPORTANT]  
+> Be careful not to fall into the pitfall of the aforementioned `uint128`
```

Comment:
I'm not clear on what this note is indicating

Reply:
It was supposed to indicate the pitfall a dev may fall into when implementing an overflow check in assembly using the patterns discussed.

Perhaps it wasn't clear :)